pkg/volume/volume_linux.go

```diff
@@ -36,60 +39,171 @@ const (
 // SetVolumeOwnership modifies the given volume to be owned by
 // fsGroup, and sets SetGid so that newly created files are owned by
 // fsGroup. If fsGroup is nil nothing is done.
-func SetVolumeOwnership(mounter Mounter, fsGroup *int64) error {
-
+func SetVolumeOwnership(mounter Mounter, fsGroup *int64, fsGroupChangePolicy *v1.PodFSGroupChangePolicy) error {
 	if fsGroup == nil {
 		return nil
 	}
 
+	if skipPermissionChange(mounter, fsGroup, fsGroupChangePolicy) {
+		klog.V(3).Infof("skipping permission and ownership change for volume %s", mounter.GetPath())
+		return nil
+	}
+
 	klog.Warningf("Setting volume ownership for %s and fsGroup set. If the volume has a lot of files then setting volume ownership could be slow, see https://github.com/kubernetes/kubernetes/issues/69699", mounter.GetPath())
 
-	return filepath.Walk(mounter.GetPath(), func(path string, info os.FileInfo, err error) error {
-		if err != nil {
+	rootDir := mounter.GetPath()
+	// We are not using filepath.Walk() because we do not want to sort files found in a directory before changing
```

Comment:
I know this is more work, but can we use the old method if the feature gate is off, and use the new method when the feature gate is enabled and they opt-into the new policy?

That way if there are problems with the bottom up approach, then we are not impacting existing workloads. As we gain more usage, then we can eventually switch over to the new method.

Reply:
restored old method if feature gate is disabled.